csharp/src/Ice/ConnectionI.cs

```diff
@@ -2643,49 +2632,93 @@ private int parseMessage(ref MessageInfo info)
         return _state == StateHolding ? SocketOperation.None : SocketOperation.Read;
     }
 
-    private void dispatchAll(InputStream stream, int requestCount, int requestId, byte compress,
-                           ServantManager servantManager, ObjectAdapter adapter)
+    private void dispatchAll(
+        InputStream stream,
+        int requestCount,
+        int requestId,
+        byte compress,
+        ObjectAdapterI adapter)
     {
-        //
-        // Note: In contrast to other private or protected methods, this
-        // operation must be called *without* the mutex locked.
-        //
+        // Note: In contrast to other private or protected methods, this method must be called *without* the mutex
+        // locked.
+
+        Object dispatcher = adapter?.dispatchPipeline;
 
-        Incoming inc = null;
         try
         {
             while (requestCount > 0)
             {
-                //
-                // Prepare the invocation.
-                //
-                bool response = !_endpoint.datagram() && requestId != 0;
-                Debug.Assert(!response || requestCount == 1);
-
-                inc = getIncoming(adapter, response, compress, requestId);
-
-                //
-                // Dispatch the invocation.
-                //
-                inc.dispatch(servantManager, stream);
+                // adapter can be null here, however the adapter set in current can't be null, and we never pass
+                // a null current.adapter to the application code. Once this file enables nullable, adapter should be
+                // adapter! below.
+                var request = new IncomingRequest(requestId, this, adapter, stream);
```

Comment:
I'm not following this comment. 

> adapter can be null here
>  Once this file enables nullable, adapter should be adapter! belo

Reply:
An outgoing connection may not have an associated object adapter, and typically doesn't have one.

On the other hand, current.adapter and by extension the IncomingRequest ctor have non-null ObjectAdapter field/parameter. 

Hence the issue and the solution described by this comment.